third_party/move/move-compiler-v2/src/function_checker.rs

```diff
@@ -56,13 +57,116 @@ pub fn check_for_function_typed_parameters(env: &mut GlobalEnv) {
+            match exp {
+                ExpData::Call(id, oper, _) => match oper {
+                    Operation::Exists(_)
+                    | Operation::BorrowGlobal(_)
+                    | Operation::MoveFrom
+                    | Operation::MoveTo
+                        if env.get_node_instantiation(*id)[0].get_struct(env).is_some() =>
+                    {
+                        let inst = env.get_node_instantiation(*id);
+                        let (mid, sid, _) = inst[0].require_struct();
+                        if mid != caller_module_id {
+                            let qualified_struct_id = mid.qualified(sid);
+                            let struct_env = env.get_struct(qualified_struct_id);
+                            access_error(
+                                env,
+                                &fun_env.get_id_loc(),
+                                id,
+                                "called",
+                                format!(
+                                    "Invalid storage operation on external type `{}`",
+                                    struct_env.get_full_name_str()
+                                ),
+                            );
+                        }
+                    },
+                    Operation::Select(mid, sid, fid) if *mid != caller_module_id => {
+                        let qualified_struct_id = mid.qualified(*sid);
+                        let struct_env = env.get_struct(qualified_struct_id);
+                        access_error(
+                            env,
+                            &fun_env.get_id_loc(),
+                            id,
+                            "accessed",
+                            format!(
+                                "Invalid access of field `{}` on external type `{}`",
+                                fid.symbol().display(struct_env.symbol_pool()),
+                                struct_env.get_full_name_str()
+                            ),
+                        );
+                    },
+                    Operation::Pack(mid, sid) => {
+                        if *mid != caller_module_id {
+                            let qualified_struct_id = mid.qualified(*sid);
+                            let struct_env = env.get_struct(qualified_struct_id);
+                            access_error(
+                                env,
+                                &fun_env.get_id_loc(),
+                                id,
+                                "packed",
+                                format!(
+                                    "Invalid pack operation on external type {}",
+                                    struct_env.get_full_name_str()
+                                ),
+                            );
+                        }
+                    },
+                    _ => {},
+                },
+                ExpData::Assign(_, pat, _) | ExpData::Block(_, pat, _, _) => {
```

Comment:
It seems like `Lambda` also has a pattern, is that a missing case?

Reply:
This check is performed after inlining. Do we still have Lambda then? cc @brmataptos

Reply:
done